apps/src/p5lab/spritelab/commands/validationCommands.js

```diff
@@ -57,10 +64,583 @@ export const commands = {
     };
   },
 
+  // Gets an array of ids for any sprite that has an associated event triggered this frame.
+  getEventSpriteIds() {
+    // We want to store any ids that are included in events logged this frame.
+    // Touch events include two distinct sprite ids.
+    let idArray = [];
+
+    // Only check for values that are new this frame.
+    for (let i = this.previous.eventLogLength; i < this.eventLog.length; i++) {
+      if (
+        // Check for each event type that includes sprite ids (NOT time or key events).
+        this.eventLog[i].includes('whenClick: ') ||
+        this.eventLog[i].includes('whileClick: ') ||
+        this.eventLog[i].includes('whenTouch: ') ||
+        this.eventLog[i].includes('whileTouch: ') ||
+        this.eventLog[i].includes('spriteCreated: ')
+      ) {
+        // Use .concat because it's possible for multiple events to be logged in the same frame.
+        idArray = idArray.concat(
+          // Take whatever was in the current eventLog entry...
+          this.eventLog[i]
+            // ...remove the spaces and create an array...
+            .split(' ')
+            // ...convert each string in the array to a number...
+            .map(Number)
+            // ...remove NaN values. (ex. 'whenClick: 0' results in [Nan, 0] above)
+            .filter(function(value) {
+              return !Number.isNaN(value);
+            })
+        );
+      }
+    }
+    return idArray;
+  },
+
+  // Used in levels (typically first frame only) to create an ordered list of success criteria
+  addCriteria(predicate, feedback) {
```

Comment:
I'm curious what `predicate` (and `feedback`) mean here.  For future visitors to the code, it could help to describe this a bit more.

Reply:
Good question. These words came from Anjali's [Blockly Validation Proposal](https://docs.google.com/document/d/1flp-BDQ9bOD6Wl2HMh0occjnVKNMM6tAsmnHZLc8I10/edit#heading=h.pqnapvd1ruox). I added a comment to describe this better.